infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,293 @@
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus.SegmentState.TRANSFERRED;
+
+/**
+ * A class responsible for managing a state transfer on the current node.
+ * It executes the state transfer for each non-transferred segment synchronously.
+ */
+@Slf4j
+@Builder
+public class StateTransferManager {
+
+    /**
+     * A data class that represents a non-empty and bounded segment to be transferred.
+     */
+    @EqualsAndHashCode
+    @Getter
+    @ToString
+    @Builder
+    @AllArgsConstructor(access = AccessLevel.PRIVATE)
+    public static class TransferSegment implements Comparable<TransferSegment> {
+        /**
+         * Start address of a segment range to transfer, inclusive and non-negative.
+         */
+        private final long startAddress;
+        /**
+         * End address of a segment range to transfer, inclusive and non-negative.
+         */
+        private final long endAddress;
+        /**
+         * A status of a transfer of a segment.
+         */
+        private final TransferSegmentStatus status;
+
+        @Override
+        public int compareTo(TransferSegment other) {
+            return Long.compare(this.getStartAddress(), other.getStartAddress());
+        }
+
+        /**
+         * Compute the total number of transferred addresses.
+         * {@link #endAddress} and {@link #startAddress} can only be non-negative longs such that
+         * {@link #endAddress} >= {@link #startAddress}.
+         *
+         * @return Sum of the total addresses transferred.
+         */
+        public long computeTotalTransferred() {
+            return endAddress - startAddress + 1L;
+        }
+
+        public static class TransferSegmentBuilder {
+
+            public void verify() {
+                if (startAddress < 0L || endAddress < 0L) {
+                    throw new IllegalStateException(
+                            String.format("Start: %s or end: %s " +
+                                    "can not be negative.", startAddress, endAddress));
+                }
+                if (startAddress > endAddress) {
+                    throw new IllegalStateException(
+                            String.format("Start: %s can not be " +
+                                    "greater than end: %s.", startAddress, endAddress));
+                }
+
+                if (status == null) {
+                    throw new IllegalStateException("Status should be defined.");
+                }
+            }
+
+            public TransferSegment build() {
+                verify();
+                return new TransferSegment(startAddress, endAddress, status);
+            }
+        }
+    }
+
+    /**
+     * A data class that represents a status of a segment to be transferred.
+     */
+    @Getter
+    @ToString
+    @EqualsAndHashCode
+    @Builder
+    public static class TransferSegmentStatus {
```

Comment:
Maybe its noteworthy to mention that this "transfer unit" should be able to fit in a single rpc.